src/pyfmi/fmi3.pyx

```diff
@@ -542,26 +563,651 @@ cdef class FMUModelBase3(FMI_BASE.ModelBase):
+        """
+        cdef int status
+        cdef np.ndarray[FMIL3.fmi3_value_reference_t, ndim=1, mode='c'] input_valueref = np.asarray(valueref, dtype = np.uint32).ravel()
+        cdef FMIL3.fmi3_string_t* val = <FMIL3.fmi3_string_t*>FMIL.malloc(sizeof(FMIL3.fmi3_string_t)*np.size(values))
+
+        if np.size(input_valueref) != np.size(values):
+            raise FMUException('The length of valueref and values are inconsistent. Note: Array variables are not yet supported')
+
+        values = [pyfmi_util.encode(item) for item in values]
+        for i in range(np.size(input_valueref)):
+            val[i] = values[i]
+
+        self._log_handler.capi_start_callback(self._max_log_size_msg_sent, self._current_log_size)
+        status = FMIL3.fmi3_import_set_string(
+            self._fmu,
+            <FMIL3.fmi3_value_reference_t*> input_valueref.data,
+            np.size(input_valueref),
+            val,
+            np.size(values)
+        )
+        self._log_handler.capi_end_callback(self._max_log_size_msg_sent, self._current_log_size)
+
+        FMIL.free(val)
+
+        if status != 0:
+            raise FMUException('Failed to set the Enum values. See the log for possibly more information.')
+
+    cpdef set_enum(self, valueref, values):
+        """
+        Sets the enum-values in the FMU as defined by the value reference(s).
+
+        Parameters::
+
+            valueref --
+                A list of value references.
+
+            values --
+                Values to be set.
+
+        Example::
+
+            model.set_enum([234, 235],[2, 10])
+
+        Calls the low-level FMI function: fmi3SetInt64
+        """
+        cdef int status
+        cdef np.ndarray[FMIL3.fmi3_value_reference_t, ndim=1, mode='c'] input_valueref = np.asarray(valueref, dtype = np.uint32).ravel()
+        cdef np.ndarray[FMIL3.fmi3_int64_t, ndim=1, mode='c'] set_value = np.asarray(values, dtype = np.int64).ravel()
+
+        if np.size(input_valueref) != np.size(set_value):
+            raise FMUException('The length of valueref and values are inconsistent. Note: Array variables are not yet supported')
+
+        self._log_handler.capi_start_callback(self._max_log_size_msg_sent, self._current_log_size)
+        status = FMIL3.fmi3_import_set_int64(
+            self._fmu,
+            <FMIL3.fmi3_value_reference_t*> input_valueref.data,
+            np.size(input_valueref),
+            <FMIL3.fmi3_int64_t*> set_value.data,
+            np.size(set_value)
+        )
+        self._log_handler.capi_end_callback(self._max_log_size_msg_sent, self._current_log_size)
+
+        if status != 0:
+            raise FMUException('Failed to set the Enum values. See the log for possibly more information.')
+
     def _get(self, variable_name):
         """
-        Helper method to get, see docstring on get.
+        Helper method to get, see docstring on get.
+        """
+        cdef FMIL3.fmi3_value_reference_t ref
+
```

Comment:
Initially I was about to suggest that we change this to a dict and map each type to a function (similar to what we did in DumpDataFMI3), however, since this is a pyx file and the set/get-methods are invoked frequently I thought that maybe we get better performance as you have done it now, so let's leave it as that.
However, after a bit of reading I learnt that for regular "def" methods within a .pyx file, type annotations help optimize operations inside the method.

I would therefore propose at least we try to type annotate the get and set methods (if possible), and perhaps we might get a tiny bit of performance boost.

Reply:
I also tried a bit; I vaguely recall that Cython supposedly reduces something like this to `switch ... case` automatically, but I couldn't see any evidence of this in the generated C code. I think this largely goes back to the question if we should be using the Enums for the FMI3_Types in the first place, if Cython can recognize these as constants.

Not sure if there is more to annotate right now?

Reply:
This is good for now. I think we can take most of it later with larger performance investigation when we are getting closer to finish this.